src/material.rs

```diff
@@ -1,45 +1,220 @@
+
+    /// Emissive texture.
+    ///
+    /// Default: `None`.
+    pub emissive_map: Option<Texture<[f32; 4]>>,
+
+    /// Metallic-roughness texture.
+    ///
+    /// Default: `None`.
+    pub metallic_roughness_map: Option<Texture<[f32; 4]>>,
+
+    /// Occlusion texture.
+    ///
+    /// Default: `None`.
+    pub occlusion_map: Option<Texture<[f32; 4]>>,
+}
+
+impl Default for Pbr {
+    fn default() -> Self {
+        Self {
+            base_color_factor: 0xFFFFFF,
+            base_color_alpha: 1.0,
+            metallic_factor: 1.0,
+            roughness_factor: 1.0,
+            occlusion_strength: 1.0,
+            emissive_factor: 0x000000,
+            normal_scale: 1.0,
+            base_color_map: None,
+            normal_map: None,
+            emissive_map: None,
+            metallic_roughness_map: None,
+            occlusion_map: None,
+        }
+    }
+}
+
+/// Parameters for a Phong reflection model.
+#[derive(Clone, Debug, Default)]
+pub struct Phong {
+    /// Solid color applied in the absense of `map`.
+    ///
+    /// Default: `0x000000` (black).
+    pub color: Color,
+
+    /// Determines the sharpness of specular highlights.
+    ///
+    /// Higher values result in sharper highlights to produce a glossy effect.
+    ///
+    /// Default: `30.0`.
+    pub glossiness: f32,
+}
+
+/// Texture for a 2D sprite.
+#[derive(Clone, Debug)]
+pub struct Sprite {
+    /// The texture the apply to the sprite.
+    pub map: Texture<[f32; 4]>,
+}
+
+/// Specifies the appearance of a [`Mesh`](struct.Mesh.html).
 #[derive(Clone, Debug)]
-pub enum Material {
-    /// Basic wireframe with specific `Color`.
-    LineBasic { color: Color },
-    /// Basic material with color, optional `Texture` and optional wireframe mode.
-    MeshBasic {
-        color: Color,
-        map: Option<Texture<[f32; 4]>>,
-        wireframe: bool,
-    },
-    /// Lambertian diffuse reflection. This technique causes all closed polygons
-    /// (such as a triangle within a 3D mesh) to reflect light equally in all
-    /// directions when rendered.
-    MeshLambert { color: Color, flat: bool },
-    /// Material that uses Phong reflection model.
-    MeshPhong { color: Color, glossiness: f32 },
-    /// Physically-based rendering material.
-    MeshPbr {
-        base_color_factor: [f32; 4],
-        metallic_roughness: [f32; 2],
-        occlusion_strength: f32,
-        emissive_factor: [f32; 3],
-        normal_scale: f32,
-
-        base_color_map: Option<Texture<[f32; 4]>>,
-        normal_map: Option<Texture<[f32; 4]>>,
-        emissive_map: Option<Texture<[f32; 4]>>,
-        metallic_roughness_map: Option<Texture<[f32; 4]>>,
-        occlusion_map: Option<Texture<[f32; 4]>>,
-    },
-    /// 2D Sprite.
-    Sprite { map: Texture<[f32; 4]> },
-    /// Custom material.
-    CustomBasicPipeline {
-        color: Color,
-        map: Option<Texture<[f32; 4]>>,
-        pipeline: BasicPipelineState,
-    },
+pub struct Material(pub(crate) Params);
```

Comment:
I think it's important for `Material` to be a `struct` since it's more accessible in the top-level documentation.

Reply:
What's wrong with `enum` accessibility?

Reply:
Enums are placed in the separate section in docs.

Reply:
^^ :+1:

Reply:
It's a minor concern though. I'll put it back for now since there's not much justification for an opaque material type right now.

Reply:
I've made it an enum again for simplicity. We can always revisit this.